resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/PreferNonAllocApiAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+using System.Linq;
+using JetBrains.Annotations;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.ExtensionsAPI.Resolve.Filters;
+using JetBrains.ReSharper.Psi.Resolve;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(IInvocationExpression), HighlightingTypes = new[] {typeof(PreferNonAllocApiWarning)})]
+    public class PreferNonAllocApiAnalyzer : UnityElementProblemAnalyzer<IInvocationExpression>
+    {
+        public PreferNonAllocApiAnalyzer([NotNull] UnityApi unityApi) : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(IInvocationExpression expression, ElementProblemAnalyzerData data,
+            IHighlightingConsumer consumer)
+        {
+            if (!(expression.InvokedExpression is IReferenceExpression referenceExpression)) 
+                return;
+
+            var reference = expression.Reference;
+            if (reference == null) 
+                return;
+            
+            var info = reference.Resolve();
+            if (info.ResolveErrorType == ResolveErrorType.OK && info.DeclaredElement is IMethod allocMethod)
+            {
+                var nonAllocMethod = GetNonAllocVersion(allocMethod, expression);
+                if (nonAllocMethod != null)
+                    consumer.AddHighlighting(new PreferNonAllocApiWarning(expression, referenceExpression, nonAllocMethod));
+            }
+        }
+
+        private static IMethod GetNonAllocVersion([NotNull]IMethod method,[NotNull] IInvocationExpression expression)
+        {
+            var originName = method.ShortName;
+            
+            string newName; // xxx[All] -> xxxNonAlloc
+            if (originName.EndsWith("All"))
+            {
+                newName = originName.Substring(0, originName.Length - 3) + "NonAlloc";
+            }
+            else
+            {
+                newName = originName + "NonAlloc";
+            }
+
+            var containingType = method.GetContainingType();
+            
+            if (containingType == null)
+                return null;
+
+            // drop out all other invocation
+            if (!containingType.GetClrName().Equals(KnownTypes.Physics) &&
+                !containingType.GetClrName().Equals(KnownTypes.Physics2D))
+                return null;
+
+            var type = TypeFactory.CreateType(containingType);
+            var table = type.GetSymbolTable(expression.PsiModule).Filter(
+                new AccessRightsFilter(new DefaultAccessContext(expression)),
+                new ExactNameFilter(newName),
+                new PredicateFilter(t => MatchSignatureAllocToNonAlloc(method, t.GetDeclaredElement() as IMethod)));
```

Comment:
Is this heuristic good enough to capture all of the methods we need? Can we rely on naming and parameter size +1? Would it be easier to just have a map of allocating to non-allocating methods?